homeassistant/components/smart_meter_texas/__init__.py

```diff
@@ -0,0 +1,132 @@
+"""The Smart Meter Texas integration."""
+import asyncio
+import logging
+
+from smart_meter_texas import Account, Client
+from smart_meter_texas.exceptions import (
+    SmartMeterTexasAPIError,
+    SmartMeterTexasAuthError,
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import aiohttp_client
+from homeassistant.helpers.update_coordinator import (
+    DataUpdateCoordinator,
+    Debouncer,
+    UpdateFailed,
+)
+
+from .const import DEBOUNCE_COOLDOWN, DOMAIN, SCAN_INTERVAL
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS = ["sensor"]
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Smart Meter Texas component."""
+    hass.data.setdefault(DOMAIN, {})
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Smart Meter Texas from a config entry."""
+
+    username = entry.data[CONF_USERNAME]
+    password = entry.data[CONF_PASSWORD]
+
+    account = Account(username, password)
+    smartmetertexas = SmartMeterTexasData(hass, entry, account)
+    try:
+        await smartmetertexas.client.authenticate()
+    except SmartMeterTexasAuthError:
+        _LOGGER.error("Username or password was not accepted")
+        return False
+    except asyncio.TimeoutError:
+        raise ConfigEntryNotReady
+
+    await smartmetertexas.setup()
+
+    async def async_update_data():
+        await smartmetertexas.read_meters()
+        return smartmetertexas
+
+    # Use a DataUpdateCoordinator to manage the updates. This is due to the
+    # Smart Meter Texas API which takes around 30 seconds to read a meter.
+    # This avoids Home Assistant from complaining about the component taking
+    # too long to update.
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name="Smart Meter Texas",
+        update_method=async_update_data,
+        update_interval=SCAN_INTERVAL,
+        request_refresh_debouncer=Debouncer(
+            hass, _LOGGER, cooldown=DEBOUNCE_COOLDOWN, immediate=True
+        ),
+    )
+
+    await coordinator.async_refresh()
```

Comment:
In testing this blocked startup for a while. `2020-08-10 02:45:42 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: smart_meter_texas
` 

Can we avoid the initial refresh here, and make the entity a RestoreEntity (example in `harmony/remote.py`)?

Reply:
Restoring state seems somewhat straightforward, but nearly everything in the sensor relies on the Meter object. What's the best way to handle that?

Reply:
This is a bit rough, but something like this should work: https://github.com/bdraco/home-assistant/commit/a8a4afb04fc97c5a22522b777af561fe7b7f4bf4

Reply:
Thanks @bdraco, that does help a lot. However it still seems to be performing a refresh at startup and therefore blocking the UI from responding.

Reply:
With the changes, startup went from 48s to 13s. The refresh shouldn't block the UI since its using `asyncio.create_task` which doesn't block startup.


In your testing was it the call to `self.account.fetch_meters` that blocked startup?

Reply:
After further testing I don't think it's blocking startup. `self.account.fetch_meters` returns very quickly.